### PR TITLE
Faster caching for text dataset

### PR DESCRIPTION
**Pull request by lhoestq**
As mentioned in #546 and #548 , hashing `data_files` contents to get the cache directory name for a text dataset can take a long time.

To make it faster I changed the hashing so that it takes into account the `path` and the `last modified timestamp` of each data file, instead of iterating through the content of each file to get a hash.

### CODE REVIEW

File: src/nlp/builder.py [hunk 1]
```diff
@@ -644,41 +643,46 @@ def _build_single_dataset(self, split: Union[str, Split], run_post_process: bool
...
+                ds._info.size_in_bytes = self.info.size_in_bytes
+                if self.info.post_processed.features is not None:
+                    if self.info.post_processed.features.type != ds.features.type:
+                        raise ValueError(
+                            "Post-processed features info don't match the dataset:\nGot\n{}\nbut expected something like\n{}".format(
+                                self.info.post_processed.features, ds.features
+                            )
+                        )
+                    else:
+                        ds.info.features = self.info.post_processed.features
```
Comment (lhoestq) [inline src/nlp/builder.py:685]: When I added the unit test to check the good caching behavior of text datasets, I found that the `DatasetInfo` were not properly updated if there was a post processing, so I fixed that

File: tests/test_search.py [hunk 2]
```diff
@@ -5,47 +5,15 @@
...
-        train_info = SplitInfo(name="train", num_examples=30)
-        test_info = SplitInfo(name="test", num_examples=30)
-        split_infos = [train_info, test_info]
-        split_dict = SplitDict()
-        split_dict.add(train_info)
-        split_dict.add(test_info)
-        info = DatasetInfo(splits=split_dict)
-        reader = ReaderTester("", info)
-        dset = Dataset(**reader.read(name, "train", split_infos))
+        dset = Dataset.from_dict({"filename": ["my_name-train" + "_" + str(x) for x in np.arange(30).tolist()]})
```
Comment (lhoestq) [inline tests/test_search.py:16]: I simplified this test too

File: src/nlp/builder.py (hunk 1 above)
Comment (thomwolf) [reply src/nlp/builder.py:685]: Good catch!

File: src/nlp/fingerprint.py [hunk 3]
```diff
@@ -89,6 +90,9 @@ def generate_fingerprint(dataset):
             continue
         hasher.update(key)
         hasher.update(state[key])
+    # hash data files last modification timestamps as well
+    for data_file in state.get("_data_files", []) + state.get("_indices_data_files", []):
+        hasher.update(os.path.getmtime(data_file["filename"]))
```
Comment (thomwolf) [inline src/nlp/fingerprint.py:95]: Nice indeed!

**Review by thomwolf — APPROVED** [review]
Awesome!

**Pull request merged**